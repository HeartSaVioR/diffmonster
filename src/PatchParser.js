export const LineType = {
  CONTEXT: ' ',
  ADDITION: '+',
  DELETION: '-',
};

export function parsePatch(patch) {
  const lines = patch.split('\n');
  const hunks = [];
  let hunk = null;
  for (var i = 0; i < lines.length; i++) {
    const line = lines[i];
    if (line.indexOf('@@ ') === 0) {
      if (hunk && hunk.lines.length > 0) {
        hunks.push(parseLines(hunk));
      }
      hunk = {
        position: i,
        lines: []
      };
    }
    hunk.lines.push(line);
  }
  if (hunk && hunk.lines.length > 0) {
    hunks.push(parseLines(hunk));
  }

  return hunks;
}

export function parseHunkHeader(header) {
  const result = /^@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@/.exec(header);
  return {
    from: {
      start: parseInt(result[1], 10),
      count: result[2] ? parseInt(result[2], 10) : 1
    },
    to: {
      start: parseInt(result[3], 10),
      count: result[4] ? parseInt(result[4], 10) : 1
    }
  };
}

function parseLines(hunk) {
  hunk.header = hunk.lines.shift();
  hunk.range = parseHunkHeader(hunk.header);
  var fromLine = hunk.range.from.start;
  var toLine = hunk.range.to.start;
  hunk.lines = hunk.lines.map((line, i) => {
    const type = line.charAt(0);
    const position = hunk.position + i;
    const content = line.substring(1);
    switch (type) {
      case LineType.CONTEXT:
        return {
          type,
          fromLine: fromLine++,
          toLine: toLine++,
          position,
          content,
        };
      case LineType.ADDITION:
        return {
          type,
          fromLine: 0,
          toLine: toLine++,
          position,
          content,
        };
      case LineType.DELETION:
        return {
          type,
          fromLine: fromLine++,
          toLine: 0,
          position,
          content,
        };
      default:
        throw new Error('invalid format');
    }
  });
  return hunk;
}
