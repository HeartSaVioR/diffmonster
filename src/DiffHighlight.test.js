import { parsePatch } from './PatchParser';
import { highlightDiff } from './DiffHighlight';

describe('DiffHighlight', () => {
  it('highlights the beginning of a line', () => {
    const patch = parsePatch(`@@ -1,3 +1,3 @@
 aaa
-bbb
+0bb
 ccc`)[0];
    expect(highlightDiff(patch)).toMatchSnapshot();
  });

  it('highlights the end of a line', () => {
    const patch = parsePatch(`@@ -1,3 +1,3 @@
 aaa
-bbb
+bb0
 ccc`)[0];
    expect(highlightDiff(patch)).toMatchSnapshot();
  });

  it('highlights the middle of a line', () => {
    const patch = parsePatch(`@@ -1,3 +1,3 @@
 aaa
-bbb
+b0b
 ccc`)[0];
    expect(highlightDiff(patch)).toMatchSnapshot();
  });

  it('does not highlight whole line', () => {
    const patch = parsePatch(`@@ -1,3 +1,3 @@
 aaa
-bbb
+000
 ccc`)[0];
    expect(highlightDiff(patch)).toMatchSnapshot();
  });

  it('does not highlight boring pairs', () => {
    const patch = parsePatch(`@@ -1,3 +1,3 @@
 aaa
-  
+     
 ccc`)[0];
    const result = highlightDiff(patch);
    expect(result.some(({ content }) => content.some(part => part.highlight))).toBe(false);
  });

  it('does not highlight mismatched hunk size', () => {
    const patch = parsePatch(`@@ -1,3 +1,3 @@
 aaa
-bbb
+b0b
+ccc`)[0];
    const result = highlightDiff(patch);
    expect(result.some(({ content }) => content.some(part => part.highlight))).toBe(false);
  });
});
