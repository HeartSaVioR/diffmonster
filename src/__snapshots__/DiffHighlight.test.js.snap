exports[`DiffHighlight does not highlight whole line 1`] = `
Array [
  Object {
    "content": Array [
      Object {
        "content": "aaa",
        "highlight": false,
      },
    ],
    "fromLine": 1,
    "position": 0,
    "toLine": 1,
    "type": " ",
  },
  Object {
    "content": Array [
      Object {
        "content": "bbb",
        "highlight": false,
      },
    ],
    "fromLine": 2,
    "position": 1,
    "toLine": 0,
    "type": "-",
  },
  Object {
    "content": Array [
      Object {
        "content": "000",
        "highlight": false,
      },
    ],
    "fromLine": 0,
    "position": 2,
    "toLine": 2,
    "type": "+",
  },
  Object {
    "content": Array [
      Object {
        "content": "ccc",
        "highlight": false,
      },
    ],
    "fromLine": 3,
    "position": 3,
    "toLine": 3,
    "type": " ",
  },
]
`;

exports[`DiffHighlight highlights the beginning of a line 1`] = `
Array [
  Object {
    "content": Array [
      Object {
        "content": "aaa",
        "highlight": false,
      },
    ],
    "fromLine": 1,
    "position": 0,
    "toLine": 1,
    "type": " ",
  },
  Object {
    "content": Array [
      Object {
        "content": "b",
        "highlight": true,
      },
      Object {
        "content": "bb",
        "highlight": false,
      },
    ],
    "fromLine": 2,
    "position": 1,
    "toLine": 0,
    "type": "-",
  },
  Object {
    "content": Array [
      Object {
        "content": "0",
        "highlight": true,
      },
      Object {
        "content": "bb",
        "highlight": false,
      },
    ],
    "fromLine": 0,
    "position": 2,
    "toLine": 2,
    "type": "+",
  },
  Object {
    "content": Array [
      Object {
        "content": "ccc",
        "highlight": false,
      },
    ],
    "fromLine": 3,
    "position": 3,
    "toLine": 3,
    "type": " ",
  },
]
`;

exports[`DiffHighlight highlights the end of a line 1`] = `
Array [
  Object {
    "content": Array [
      Object {
        "content": "aaa",
        "highlight": false,
      },
    ],
    "fromLine": 1,
    "position": 0,
    "toLine": 1,
    "type": " ",
  },
  Object {
    "content": Array [
      Object {
        "content": "bb",
        "highlight": false,
      },
      Object {
        "content": "b",
        "highlight": true,
      },
    ],
    "fromLine": 2,
    "position": 1,
    "toLine": 0,
    "type": "-",
  },
  Object {
    "content": Array [
      Object {
        "content": "bb",
        "highlight": false,
      },
      Object {
        "content": "0",
        "highlight": true,
      },
    ],
    "fromLine": 0,
    "position": 2,
    "toLine": 2,
    "type": "+",
  },
  Object {
    "content": Array [
      Object {
        "content": "ccc",
        "highlight": false,
      },
    ],
    "fromLine": 3,
    "position": 3,
    "toLine": 3,
    "type": " ",
  },
]
`;

exports[`DiffHighlight highlights the middle of a line 1`] = `
Array [
  Object {
    "content": Array [
      Object {
        "content": "aaa",
        "highlight": false,
      },
    ],
    "fromLine": 1,
    "position": 0,
    "toLine": 1,
    "type": " ",
  },
  Object {
    "content": Array [
      Object {
        "content": "b",
        "highlight": false,
      },
      Object {
        "content": "b",
        "highlight": true,
      },
      Object {
        "content": "b",
        "highlight": false,
      },
    ],
    "fromLine": 2,
    "position": 1,
    "toLine": 0,
    "type": "-",
  },
  Object {
    "content": Array [
      Object {
        "content": "b",
        "highlight": false,
      },
      Object {
        "content": "0",
        "highlight": true,
      },
      Object {
        "content": "b",
        "highlight": false,
      },
    ],
    "fromLine": 0,
    "position": 2,
    "toLine": 2,
    "type": "+",
  },
  Object {
    "content": Array [
      Object {
        "content": "ccc",
        "highlight": false,
      },
    ],
    "fromLine": 3,
    "position": 3,
    "toLine": 3,
    "type": " ",
  },
]
`;
